flytekit/types/structured/structured_dataset.py

```diff
@@ -0,0 +1,380 @@
+    """
+
+    _SUPPORTED_TYPES: typing.Dict[Type, LiteralType] = {
+        _np.int32: type_models.LiteralType(simple=type_models.SimpleType.INTEGER),
+        _np.int64: type_models.LiteralType(simple=type_models.SimpleType.INTEGER),
+        _np.uint32: type_models.LiteralType(simple=type_models.SimpleType.INTEGER),
+        _np.uint64: type_models.LiteralType(simple=type_models.SimpleType.INTEGER),
+        int: type_models.LiteralType(simple=type_models.SimpleType.INTEGER),
+        _np.float32: type_models.LiteralType(simple=type_models.SimpleType.FLOAT),
+        _np.float64: type_models.LiteralType(simple=type_models.SimpleType.FLOAT),
+        float: type_models.LiteralType(simple=type_models.SimpleType.FLOAT),
+        _np.bool_: type_models.LiteralType(simple=type_models.SimpleType.BOOLEAN),  # type: ignore
+        bool: type_models.LiteralType(simple=type_models.SimpleType.BOOLEAN),
+        _np.datetime64: type_models.LiteralType(simple=type_models.SimpleType.DATETIME),
+        _datetime.datetime: type_models.LiteralType(simple=type_models.SimpleType.DATETIME),
+        _np.timedelta64: type_models.LiteralType(simple=type_models.SimpleType.DURATION),
+        _datetime.timedelta: type_models.LiteralType(simple=type_models.SimpleType.DURATION),
+        _np.string_: type_models.LiteralType(simple=type_models.SimpleType.STRING),
+        _np.str_: type_models.LiteralType(simple=type_models.SimpleType.STRING),
+        _np.object_: type_models.LiteralType(simple=type_models.SimpleType.STRING),
+        str: type_models.LiteralType(simple=type_models.SimpleType.STRING),
+    }
+
+    DATASET_DECODING_HANDLERS: Dict[Type[Any], Dict[Type[Any], DatasetDecodingHandler]] = {}
+    DATASET_ENCODING_HANDLERS: Dict[Type[Any], Dict[Type[Any], DatasetEncodingHandler]] = {}
+    DATASET_PERSISTENCE_HANDLERS: Dict[Type[typing.Any], Dict[Type[Any], DatasetPersistenceHandler]] = {}
+    DATASET_RETRIEVAL_HANDLERS: Dict[Type[Any], Dict[Type[Any], DatasetRetrievalHandler]] = {}
+    Handlers = typing.Union[
+        DatasetDecodingHandler, DatasetEncodingHandler, DatasetPersistenceHandler, DatasetRetrievalHandler
+    ]
+    _REGISTER_TYPES: List[Type] = []
+
+    def __init__(self):
+        super().__init__("StructuredDataset Transformer", StructuredDataset)
+
+    def _get_dataset_column_literal_type(self, t: Type):
+        if t in self._SUPPORTED_TYPES:
+            return self._SUPPORTED_TYPES[t]
+        if hasattr(t, "__origin__") and t.__origin__ == list:
+            return type_models.LiteralType(collection_type=self._get_dataset_column_literal_type(t.__args__[0]))
+        if hasattr(t, "__origin__") and t.__origin__ == dict:
+            return type_models.LiteralType(map_value_type=self._get_dataset_column_literal_type(t.__args__[1]))
+        raise AssertionError(f"type {t} is currently not supported by StructuredDataset")
+
+    def _get_dataset_type(self, t: Type[StructuredDataset]) -> StructuredDatasetType:
```

Comment:
The type signature here isn't right though.  It should be `Union[Type[StructuredDataset], typing.Any]`.  The goal of this function, and of the caller (`get_literal_type`) is that it takes in a Python type, and returns the Flyte type.  The python type can be a `StructuredDataset` (with or without columns, (or perhaps Annotated)) but it can also be a `pd.DataFrame` or a `some_future_lib.DataFrame`.  The idea here is that the translation from the Python type to the Flyte type needs to be redirected to the registered handlers somehow - this is the part that I didn't think through.

But like,
```
TypeEngine.to_literal_type(my_custom.DataFrame)
```
should fail with the stock flytekit.  But after the user calls
```
FLYTE_DATASET_TRANSFORMER.register_handler(MyCustomEncodingHandler)
# and, or
FLYTE_DATASET_TRANSFORMER.register_handler(MyCustomPersistenceHandler)
```
then the call to the type engine should succeed.

Reply:
For custom dataframe, I think we have a few ways to translate it to FlyteType.
1. Users must use StructuredDataset with the column in the input/output signature.
2. Users must tell us how to convert the dataframe to arrow or parquet, and then we can convert arrow or parquet data column type to Flyte Type
3. Users implement `to_literal_type` by themself